File: 726-number-of-atoms/number-of-atoms.cpp
class Solution {
public:
    bool isDigit(char a) {
        return a >= '0' && a <= '9';
    }

    bool isLetter(char a) {
        return (a >= 'A' && a <= 'Z') || (a >= 'a' && a <= 'z');
    }

    string countOfAtoms(string formula) {
        map<string, int> atomCount;
        stack<pair<string, int>> st;
        int n = formula.size();
        int i = 0;

        while (i < n) {
            if (formula[i] == '(') {
                st.push({"(", 0});
                i++;
            } else if (formula[i] == ')') {
                i++;
                int count = 0;
                while (i < n && isDigit(formula[i])) {
                    count = count * 10 + (formula[i] - '0');
                    i++;
                }
                count = (count == 0) ? 1 : count;
                stack<pair<string, int>> tempStack;
                while (!st.empty() && st.top().first != "(") {
                    auto top = st.top();
                    st.pop();
                    top.second *= count;
                    tempStack.push(top);
                }
                st.pop();  // remove '('
                while (!tempStack.empty()) {
                    st.push(tempStack.top());
                    tempStack.pop();
                }
            } else {
                string atom;
                atom += formula[i++];
                if (i < n && formula[i] >= 'a' && formula[i] <= 'z') {
                    atom += formula[i++];
                }
                int count = 0;
                while (i < n && isDigit(formula[i])) {
                    count = count * 10 + (formula[i] - '0');
                    i++;
                }
                count = (count == 0) ? 1 : count;
                st.push({atom, count});
            }
        }

        while (!st.empty()) {
            auto top = st.top();
            st.pop();
            atomCount[top.first] += top.second;
        }

        string result;
        for (const auto &pair : atomCount) {
            result += pair.first;
            if (pair.second > 1) {
                result += to_string(pair.second);
            }
        }

        return result;
    }
};

File: 726-number-of-atoms/number-of-atoms.java
class Solution {
   public String countOfAtoms(String formula) {
        Stack<Map<String, Integer>> stack = new Stack<>();
        stack.push(new HashMap<>());
        int n = formula.length();
        int i = 0;

        while (i < n) {
            char c = formula.charAt(i);
            if (c == '(') {
                stack.push(new HashMap<>());
                i++;
            } else if (c == ')') {
                Map<String, Integer> top = stack.pop();
                i++;
                String countStr="";
                while (i < n && Character.isDigit(formula.charAt(i))) {
                    countStr+=formula.charAt(i);
                    i++;
                }
                int count =countStr==""? 1:Integer.parseInt(countStr);
                Map<String, Integer> current = stack.peek();
                for (String atom : top.keySet()) {
                    current.put(atom, current.getOrDefault(atom, 0) + top.get(atom) * count);
                }
            } else {
                int start = i;
                i++;
                while (i < n && Character.isLowerCase(formula.charAt(i))) {
                   // atom+=formula.charAt(i);
                    i++;
                }
                String atom = formula.substring(start, i);
                  String countStr="";
                while (i < n && Character.isDigit(formula.charAt(i))) {
                    countStr+=formula.charAt(i);
                    i++;
                }
                int count =countStr==""? 1:Integer.parseInt(countStr);
              
                Map<String, Integer> current = stack.peek();
                current.put(atom, current.getOrDefault(atom, 0) + count);
            }
        }

        Map<String, Integer> countMap = stack.pop();
        TreeMap<String, Integer> sortedMap = new TreeMap<>(countMap);
        StringBuilder result = new StringBuilder();
        for (String atom : sortedMap.keySet()) {
            result.append(atom);
            int count = sortedMap.get(atom);
            if (count > 1) {
                result.append(count);
            }
        }
        return result.toString();
    }
}